test/integration_test.go

```diff
@@ -1170,6 +1171,19 @@ func TestAutoRetryExhaustRetries(t *testing.T) {
 	assert.NotContains(t, out.String(), "Apply complete!")
 }
 
+func TestAutoRetryCustomRetryableErrors(t *testing.T) {
+	t.Parallel()
+
+	out := new(bytes.Buffer)
+	rootPath := copyEnvironment(t, TEST_FIXTURE_AUTO_RETRY_CUSTOM_ERRORS)
+	modulePath := util.JoinPath(rootPath, TEST_FIXTURE_AUTO_RETRY_CUSTOM_ERRORS)
+	err := runTerragruntCommand(t, fmt.Sprintf("terragrunt apply --auto-approve --terragrunt-non-interactive --terragrunt-working-dir %s", modulePath), out, os.Stderr)
+
+	assert.NotNil(t, err)
+	assert.Contains(t, out.String(), "My own little error")
+	assert.NotContains(t, out.String(), "Apply complete!")
```

Comment:
This checks that the `apply` eventually failed, but how do we know this actually did some retries? 

You probably want to test a fixture that fails the first N items but then succeeds on try N+1. Here's one approach for doing that:

1. Update the Terraform test fixture to:

    1. Take in a `url` variable.
    1. Use the [`http`](https://www.terraform.io/docs/providers/http/data_source.html) data source to make an HTTP request to `url`.
    1. Update the Bash script in the `null_resource` to exit with an error if the response the `http` data source gets is not a 200 OK.
    1.  Add `triggers { run_always = timestamp() }` to the `null_resource` so it runs every time you run `apply`, and not just on the very first time.

1. Update the integration test to:

    1. Fire up a [local HTTP server](https://golang.org/pkg/net/http/) listening on some port.
    1. Configure that HTTP server to return a 500 error on the first N requests, but a 200 OK on request N+1.
    1. Run `terragrunt apply -var url=http://localhost:<PORT>`, make sure the logs mention errors/retries, but that the `apply` ultimately succeeds.

Let me know if that makes sense!

Reply:
It makes sense but I wonder if firing up an HTTP server is a bit of an overkill - maybe we can do it by keeping the state in a file in `/tmp` (like we discussed this morning). What do you think?

Reply:
Yup, you could use a file-based approach too, but it might be a bit more error prone as files have to be cleaned up so that the outputs from previous test runs don't cause issues with subsequent test runs. To do it with files, you'd do something like this:

1. Use the [`random_id`](https://www.terraform.io/docs/providers/random/r/id.html) resource to generate a unique filename on the very first `apply`.

1. Update the Bash script in the `null_resource` to: 

    1. Check if the file with the name from `random_id` exists in `/tmp`. 
    1. If it does not, write it to disk, and exit with an error. 
    1. If it does, exit without an error.

1. Add `triggers { run_always = timestamp() }` to the `null_resource` so it runs every time you run `apply`, and not just on the very first time.

This way, the module will fail the first time you run `apply`, but succeed after that.

Reply:
Awesome - done!